redirect_path patch for restoring cpt

studyztp opened the pull request:
Modify the FDArray::unserialize function to perform a checkPathRedirect if a Process pointer is passed in.
Currently when restoring a checkpoint, it doesn't perform checkPathRedirect for files that were opened during checkpointing. This patch adds a checkPathRedirect in the FDArray::unserialize to redirect app path for restoring checkpoints.

studyztp committed:
sim: check redirect path when unserialize for cpt

Change-Id: I55b8ce1770b0580d52b8dfa782572d492c1bf727

studyztp committed:
sim: modifed the type of path

Change-Id: I56be3b62b1804371b9b9e0f84ee1ec49cbedf553

powerjg commented on src/sim/fd_array.cc:
```diff
@@ -367,7 +368,7 @@ FDArray::serialize(CheckpointOut &cp) const {
 
 void
-FDArray::unserialize(CheckpointIn &cp) {
+FDArray::unserialize(CheckpointIn &cp, SimObject* process_ptr) {
```
Shouldn't this be `Process*`?

studyztp replied:
Fixed it.

powerjg commented on src/sim/fd_array.hh:
```diff
@@ -36,9 +36,9 @@
-
 #include "sim/fd_entry.hh"
 #include "sim/serialize.hh"
+#include "sim/sim_object.hh"
```
Don't include this.

Instead, inside the `gem5` namespace add the line `class Process;` This is a "forward declaration"

studyztp replied:
I added it. Now I know what is "forward declaration". Thank you!

powerjg commented on src/sim/fd_array.hh:
```diff
@@ -117,7 +117,11 @@ class FDArray : public Serializable
      */
     void serialize(CheckpointOut &cp) const override;
-    void unserialize(CheckpointIn &cp) override;
+    void unserialize(CheckpointIn &cp, SimObject* process_ptr );
```
Should be `Process *`

studyztp replied:
Fixed it.

BobbyRBruce commented on src/sim/fd_array.hh:
```diff
@@ -36,9 +36,9 @@
 #include <map>
 #include <memory>
 #include <string>
-
```
There should be a new line here. Please re-add it.

studyztp replied:
I re-added it. Thanks for reviewing this!

BobbyRBruce reviewed (changes requested):
Assuming Jason's concerns are addressed, everything else bar a small line deletion is fine with me.

studyztp committed:
sim: check redirect path when unserialize for cpt

sim/fd_array.hh:
Add "class Process;" to forward declare Process for unserialize
function to pass in a Process object pointer.
Fix the styling issue with include files.

sim/fd_array.cc"
Add comments.

Change-Id: Ifb21eb1c7bad119028b8fd8e610a125100fde696

BobbyRBruce reviewed (dismissed):
Fine by me but I'll let @powerjg  sign off.

powerjg commented on src/sim/fd_array.cc:
```diff
@@ -418,11 +419,26 @@ FDArray::unserialize(CheckpointIn &cp) {
+        std::string path;
+
+        if (process_ptr)
+        {
```
These curly braces should be on the previous line

powerjg commented on src/sim/fd_array.cc:
```diff
@@ -418,11 +419,26 @@ FDArray::unserialize(CheckpointIn &cp) {
+            path = process_ptr->checkPathRedirect(this_ffd->getFileName());
+        }
+        else
+        {
```
Also this one

powerjg reviewed (changes requested):
One more tiny style thing

studyztp committed:
sim:fixed some style issues

Change-Id: I0832a8b68e802e9671b755d3a71fd9c8f17e1648

studyztp commented:
> One more tiny style thing

Fixed them. Thanks!
